dali/operators/decoder/video/video_decoder_base.h

```diff
@@ -15,121 +15,330 @@
+    } else if (pad_mode_str == "constant") {
+      boundary_type_ = boundary::BoundaryType::CONSTANT;
+    } else if (pad_mode_str == "edge") {
+      boundary_type_ = boundary::BoundaryType::CLAMP;
+    } else if (pad_mode_str == "reflect_1001" || pad_mode_str == "symmetric") {
+      boundary_type_ = boundary::BoundaryType::REFLECT_1001;
+    } else if (pad_mode_str == "reflect_101" || pad_mode_str == "reflect") {
+      boundary_type_ = boundary::BoundaryType::REFLECT_101;
+    } else {
+      DALI_FAIL(make_string("Invalid pad_mode: ", pad_mode_str, "\n",
+                            "Valid options are: none, constant, edge, reflect_1001, reflect_101"));
+    }
+
+    if (boundary_type_ == boundary::BoundaryType::CONSTANT) {
+      pad_value_ = spec_.template GetArgument<int>("pad_value");
+      DALI_ENFORCE(pad_value_ >= 0 && pad_value_ <= 255, "pad_value must be in range [0, 255]");
+    }
+  }
 
   DISABLE_COPY_MOVE_ASSIGN(VideoDecoderBase);
 
  protected:
-  void ValidateInput(const Workspace &ws) {
+  std::unique_ptr<FramesDecoderImpl> CreateDecoder(const char *data, size_t size, bool build_index,
+                                                   std::string_view source_info,
+                                                   cudaStream_t stream = 0) {
+    if constexpr (std::is_same_v<Backend, CPUBackend>) {
+      return std::make_unique<FramesDecoderImpl>(data, size, build_index, true, -1, source_info);
+    } else {
+      return std::make_unique<FramesDecoderImpl>(data, size, stream, build_index, -1, source_info);
+    }
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) override {
     const auto &input = ws.Input<InBackend>(0);
-    DALI_ENFORCE(input.type() == DALI_UINT8,
-                 "Type of the input buffer must be uint8.");
-    DALI_ENFORCE(input.sample_dim() == 1,
-                 "Input buffer must be 1-dimensional.");
+
+    // Validate input
+    DALI_ENFORCE(input.type() == DALI_UINT8, "Type of the input buffer must be uint8.");
+    DALI_ENFORCE(input.sample_dim() == 1, "Input buffer must be 1-dimensional.");
     for (int64_t i = 0; i < input.num_samples(); ++i) {
       DALI_ENFORCE(input[i].shape().num_elements() > 0,
                    make_string("Incorrect sample at position: ", i, ". ",
                                "Video decoder does not support empty input samples."));
     }
+
+    int batch_size = input.num_samples();
+
+    // Get frame selection parameters
+    if (start_frame_.HasValue())
+      start_frame_.Acquire(spec_, ws, batch_size);
+    if (stride_.HasValue())
+      stride_.Acquire(spec_, ws, batch_size);
+    if (sequence_length_.HasValue())
+      sequence_length_.Acquire(spec_, ws, batch_size);
+    if (frames_.HasValue())
+      frames_.Acquire(spec_, ws, batch_size);
+
+    frames_decoders_.resize(batch_size);
+
+    // Build index if we need to seek
+    bool build_index = start_frame_.HasValue() || frames_.HasValue();
```

Comment:
Maybe we can decide per each decoder instance if the `start_frame` is `!=0` or `frames_[0]  != 0`? Or it is too early to optimize this?